pkg/etcdcli/etcdcli_pool.go

```diff
@@ -0,0 +1,136 @@
+package etcdcli
+
+import (
+	"context"
+	"fmt"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/klog/v2"
+	"reflect"
+	"sort"
+	"sync"
+	"time"
+)
+
+// etcdClientPool fulfills these requirements:
+// * cache clients to avoid re-creating them all the time (TLS handshakes are expensive after all)
+// * return an exclusively unused client, no other can acquire the same client at that time
+// * health checking a client before using it (using list), return a new one if unhealthy and closing the old one
+// * update endpoints, to be always up to date with the changes
+// * return a used client to the pool, making it available to consume again
+type etcdClientPool struct {
+	pool          *sync.Pool
```

Comment:
some learnings from sync.Pool:
* it's really fast
* it does not actually work like a queue, when you return an item there is no guarantee to get it back during the next invocation (meaning it would create a new client)
* matter of fact, load testing this with the readyz endpoint is actually causing it to create gazillions of clients without actually closing the ones that it removed.


So with all the above, I think sync.Pool will also create a leak, will be a bit faster though. Will checkout another way to implement this tomorrow.

Reply:
@tjungblu  great comments thanks 👍🏽 🙏🏽 

I actually liked the approach, what are other options ? 

I think RWMutex is not a one, it was implemented like this right ?

Reply:
yeah I thought the sync.Pool would take all the synchronization pain away as well :/
we could wrap this into a single mutex backed with an array (as a queue). I think go channels should also work for that.